stan/math/rev/core/var.hpp

```diff
@@ -385,44 +629,91 @@ class var_value {
+  inline auto coeff(Eigen::Index i) {
+    using vari_sub = decltype(vi_->coeff(i));
+    vari_sub* vari_coeff = new vari_sub(vi_->coeff(i));
+    reverse_pass_callback([this_vi = this->vi_, vari_coeff, i]() {
+      auto tmp_adj = this_vi->adj_.coeffRef(i);
+      this_vi->adj_.coeffRef(i) += vari_coeff->adj_;
+      vari_coeff->adj_ += tmp_adj;
     });
-    return var_value<double>(vari_coeff);
+    return var_value<typename vari_sub::value_type>(vari_coeff);
   }
 
   /**
    * View element of eigen matrices
    * @param i Row to access
    * @param j Column to access
    */
-  inline auto coeff(Eigen::Index i, Eigen::Index j) const {
-    using vari_coeff_type = decltype(vi_->coeff(i, j));
-    auto* vari_coeff = new vari_coeff_type(vi_->coeff(i, j));
-    reverse_pass_callback([vari_coeff, this, i, j]() mutable {
-      this->vi_->adj_(i, j) += vari_coeff->adj_;
+  inline auto coeff(Eigen::Index i, Eigen::Index j) {
+    using vari_sub = decltype(vi_->coeff(i, j));
+    vari_sub* vari_coeff = new vari_sub(vi_->coeff(i, j));
+    reverse_pass_callback([this_vi = this->vi_, vari_coeff, i, j]() {
+      auto tmp_adj = this_vi->adj_.coeffRef(i, j);
+      this_vi->adj_.coeffRef(i, j) += vari_coeff->adj_;
+      vari_coeff->adj_ += tmp_adj;
     });
-    return var_value<double>(vari_coeff);
+    return var_value<typename vari_sub::value_type>(vari_coeff);
+  }
+
+  /**
+   * View element of eigen matrices
+   * @param i Element to access
+   */
+  inline auto operator()(Eigen::Index i) { return this->coeff(i); }
+
+  /**
+   * View element of eigen matrices
+   * @param i Row to access
+   * @param j Column to access
+   */
+  inline auto operator()(Eigen::Index i, Eigen::Index j) {
+    return this->coeff(i, j);
   }
 
   /**
    * View element of eigen matrices
    * @param i Element to access
    */
-  inline auto operator()(Eigen::Index i) const { return this->coeff(i); }
+  inline auto coeffRef(Eigen::Index i) { return this->coeff(i); }
 
   /**
    * View element of eigen matrices
    * @param i Row to access
    * @param j Column to access
    */
-  inline auto operator()(Eigen::Index i, Eigen::Index j) const {
+  inline auto coeffRef(Eigen::Index i, Eigen::Index j) {
     return this->coeff(i, j);
   }
 
+  /**
+   * Return an expression that operates on the rows of the matrix `vari`
+   */
+  inline auto rowwise() {
+    using vari_sub = decltype(vi_->rowwise());
+    using var_sub = var_value<typename vari_sub::value_type>;
+    return var_sub(new vari_sub(vi_->rowwise()));
+  }
+
+  /**
+   * Return an expression that operates on the columns of the matrix `vari`
+   */
+  inline auto colwise() {
+    using vari_sub = decltype(vi_->colwise());
+    using var_sub = var_value<typename vari_sub::value_type>;
+    return var_sub(new vari_sub(vi_->colwise()));
+  }
+
+  /**
+   * Return an expression an expression to reverse the order of the coefficients
+   * inside of a `vari` matrix
+   */
+  inline auto reverse() {
```

Comment:
Should this require a vector or can we do it on a matrix?

Reply:
We can do it on matrices

Reply:
I see how it can be done on rowwise or colwise view into matrix. But what does it do on a plain matrix?

Reply:
It's pretty much like running `rowwise().reverse().colwise().reverse()`

https://godbolt.org/z/deeo8d